src/object/object_primitive.c

```diff
@@ -13750,15 +13964,54 @@ mr_readval_varnchar_internal (OR_BUF * buf, DB_VALUE * value, TP_DOMAIN * domain
       /* Branch according to convention based on size */
       if (!copy)
 	{
-	  str_length = or_get_varchar_length (buf, &rc);
 	  if (TP_DOMAIN_COLLATION_FLAG (domain) != TP_DOMAIN_COLL_NORMAL)
 	    {
 	      assert (false);
 	      return ER_FAILED;
 	    }
-	  db_make_varnchar (value, precision, buf->ptr, str_length, TP_DOMAIN_CODESET (domain),
-			    TP_DOMAIN_COLLATION (domain));
-	  value->need_clear = false;
+
+	  /* Get the decompressed and compressed size of the string stored in buffer.
+	     If compressed_size is set to -1, then the string is not compressed at all.
+	     If compressed_size is set to 0, the string was attempted to be compressed, however it failed.
+	   */
+	  rc = or_get_varchar_compression_lengths (buf, &compressed_size, &decompressed_size);
+	  if (rc != NO_ERROR)
+	    {
+	      return rc;
+	    }
+	  if (compressed_size > 0)
+	    {
+	      str_length = compressed_size;
+	      string = db_private_alloc (NULL, decompressed_size + 1);
+	      if (string == NULL)
+		{
+		  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1,
+			  decompressed_size * sizeof (char));
+		  return rc;
+		}
+
+	      /* Getting data from the buffer. */
+	      rc = do_get_compressed_data_from_buffer (buf, string, compressed_size, decompressed_size);
+
+	      if (rc != NO_ERROR)
+		{
+		  return rc;
+		}
+
+	      string[decompressed_size] = '\0';
```

Comment:
Putting a nul terminator in 'string' is not necessary here. Function 'do_get_compressed_data_from_buffer' already does this.